src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -0,0 +1,275 @@
+            }
+
+            public void Dispose()
+            {
+                lock (_gate)
+                {
+                    if (_disposed)
+                    {
+                        return;
+                    }
+
+                    _disposed = true;
+                }
+
+                _fileChangeWatcher.EnqueueWork(
+                    service =>
+                    {
+                        // Since we put all of our work in a queue, we know that if we had tried to advise file or directory changes,
+                        // it must have happened before now
+                        if (_directoryFilePathOpt != null)
+                        {
+                            ErrorHandler.ThrowOnFailure(service.UnadviseDirChange(_directoryWatchCookie));
+                        }
+
+                        foreach (var token in _activeFileWatchingTokens)
+                        {
+                            ErrorHandler.ThrowOnFailure(service.UnadviseFileChange(token.Cookie.Value));
+                        }
+                    });
+            }
+
+            public IFileWatchingToken EnqueueWatchingFile(string filePath)
+            {
+                // If we already have this file under our path, we don't have to do additional watching
+                if (_directoryFilePathOpt != null && filePath.StartsWith(_directoryFilePathOpt))
+                {
+                    return _noOpFileWatchingToken;
+                }
+
+                var token = new FileWatchingToken();
+
+                lock (_gate)
+                {
+                    _activeFileWatchingTokens.Add(token);
+                }
+
+                _fileChangeWatcher.EnqueueWork(service =>
+                {
+                    uint cookie;
+                    ErrorHandler.ThrowOnFailure(service.AdviseFileChange(filePath, (uint)(_VSFILECHANGEFLAGS.VSFILECHG_Size | _VSFILECHANGEFLAGS.VSFILECHG_Time), this, out cookie));
+
+                    lock (_gate)
+                    {
+                        token.Cookie = cookie;
+                    }
+                });
+
+                return token;
+            }
+
+            public void StopWatchingFile(IFileWatchingToken token)
+            {
+                var typedToken = token as FileWatchingToken;
+
+                if (typedToken == null)
+                {
+                    throw new ArgumentException("The token passed did not originate from this service.", nameof(token));
+                }
+
+                if (typedToken == _noOpFileWatchingToken)
+                {
+                    // This file never required a direct file watch, our main subscription covered it.
+                    return;
+                }
+
+                lock (_gate)
+                {
+                    if (!_activeFileWatchingTokens.Remove(typedToken))
+                    {
+                        throw new InvalidOperationException("This token was no longer being watched.");
```

Comment:
>throw new InvalidOperationException("This token was no longer being watched."); [](start = 24, length = 79)

Contract.Throw since this is not public API.